consensus/leader.go

```diff
@@ -55,21 +61,20 @@ func (consensus *Consensus) announce(block *types.Block) {
 	consensus.FBFTLog.AddBlock(block)
 
 	// Leader sign the block hash itself
-	consensus.Decider.SubmitVote(
-		quorum.Prepare,
-		consensus.PubKey,
-		consensus.priKey.SignHash(consensus.blockHash[:]),
-		common.BytesToHash(consensus.blockHash[:]),
-	)
-	if err := consensus.prepareBitmap.SetKey(
-		consensus.PubKey, true,
-	); err != nil {
-		consensus.getLogger().Warn().Err(err).Msg(
-			"[Announce] Leader prepareBitmap SetKey failed",
+	for i, key := range consensus.PubKey.PublicKey {
+		consensus.Decider.SubmitVote(
+			quorum.Prepare,
+			key,
+			consensus.priKey.PrivateKey[i].SignHash(consensus.blockHash[:]),
```

Comment:
how can we make sure the consensus.PubKey.PublicKey and consensus.priKey.PrivateKey are consistent in terms of orders? Better refactor them into a class? Or just always derive pubkey from priKey.

Reply:
when we read the multi-bls keys, we append private key and corresponding public key at once, hence order is unlikely to mess-up unless `GetPublicKey` from private key that is being loaded panics. after loading, the struct is never modified and only read, hence it is safe to assume that indices based access will always be consistent. Further, it may not be a good idea to derive pub key from private every time, as it an expensive c call to bls library. do you agree with my reasoning?

Reply:
I agree with the expensive part. But assuming the order is fine may not be good. Let's add a TODO to enforce the order by making pri/pub key a tuple.

Reply:
also, consensus does not expose the priKey, only PubKey is exposed. Making them into a tuple may leak information?